diff --git a/src/cadical.hpp b/src/cadical.hpp
index 7a581b32b7..1cb519e828 100644
--- a/src/cadical.hpp
+++ b/src/cadical.hpp
@@ -1053,6 +1053,8 @@
   //
   static void build (FILE *file, const char *prefix = "c ");
 
+#include "cadical_extension.hpp"
+
 private:
   //==== start of state ====================================================
 
diff --git a/src/ccadical.cpp b/src/ccadical.cpp
index 377e505b0e..8c438c67f7 100644
--- a/src/ccadical.cpp
+++ b/src/ccadical.cpp
@@ -206,3 +206,5 @@
   ((Wrapper *) wrapper)->solver->unphase (lit);
 }
 }
+
+#include "ccadical_extension.cpp"
diff --git a/src/ccadical.h b/src/ccadical.h
index 256bd1cb54..c1af560e1f 100644
--- a/src/ccadical.h
+++ b/src/ccadical.h
@@ -65,6 +65,8 @@
 #define ccadical_sat ccadical_solve
 #define ccadical_deref ccadical_val
 
+#include "ccadical_extension.h"
+
 /*------------------------------------------------------------------------*/
 #ifdef __cplusplus
 }
diff --git a/src/congruence.cpp b/src/congruence.cpp
index a3fc19ea4b..2d2afe2afa 100644
--- a/src/congruence.cpp
+++ b/src/congruence.cpp
@@ -4663,14 +4663,15 @@
   }
   goccs (src).clear ();
 
-#ifndef NDEBUG
-  for (const auto &occs : gtab) {
-    for (auto g : occs) {
-      assert (g);
-      assert (g->garbage || !gate_contains (g, src));
-    }
-  }
-#endif
+  // CJ: these debug checks massively slow down tests
+  // #ifndef NDEBUG
+  //   for (const auto &occs : gtab) {
+  //     for (auto g : occs) {
+  //       assert (g);
+  //       assert (g->garbage || !gate_contains (g, src));
+  //     }
+  //   }
+  // #endif
   assert (lrat_chain.empty ());
   return true;
 }
diff --git a/src/congruence.hpp b/src/congruence.hpp
index b7797e2e58..391dbddbb4 100644
--- a/src/congruence.hpp
+++ b/src/congruence.hpp
@@ -278,6 +278,11 @@
   bool operator== (Gate const &lhs) {
     return tag == lhs.tag && rhs == lhs.rhs;
   }
+  // backport of
+  // https://github.com/arminbiere/cadical/commit/ba439a485e39097d27a1c933f9ed9b8099b52969
+  Gate ()
+      : lhs (0), garbage (false), indexed (false), marked (false),
+        shrunken (false), neg_lhs_ids () {}
 };
 
 typedef vector<Gate *> GOccs;
diff --git a/src/solver.cpp b/src/solver.cpp
index b3f271cb07..5cca2fbc8d 100644
--- a/src/solver.cpp
+++ b/src/solver.cpp
@@ -1856,8 +1856,12 @@
   if (!strcmp (opt, "eliminated"))
     return internal->stats.all.eliminated +
            internal->stats.all.fasteliminated;
-  if (!strcmp (opt, "subsitutued"))
+  // backport of
+  // https://github.com/arminbiere/cadical/commit/095bcf986bde3b147190994c4a71f4af25322630
+  if (!strcmp (opt, "substituted"))
     return internal->stats.all.substituted;
   return -1;
 }
 } // namespace CaDiCaL
+
+#include "solver_extension.cpp"
